presto-main/src/main/java/io/prestosql/sql/planner/optimizations/PredicatePushDown.java

```diff
@@ -504,13 +509,50 @@ public PlanNode visitJoin(JoinNode node, RewriteContext<Expression> context)
                             (!newJoinFilter.isPresent() || areExpressionsEquivalent(newJoinFilter.get(), node.getFilter().get()));
 
+            Map<String, Symbol> dynamicFilters = ImmutableMap.of();
+            List<Expression> predicates = ImmutableList.of();
+            if (node.getType() == INNER && isEnableDynamicFiltering(session)) {
+                // New equiJoinClauses could potentially not contain symbols used in current dynamic filters.
+                // Since we use PredicatePushdown to push dynamic filters themselves,
+                // instead of separate ApplyDynamicFilters rule we derive dynamic filters within PredicatePushdown itself.
+                // Even if equiJoinClauses match, current dynamic filters may not match equiJoinClauses
+                Set<Symbol> equiJoinClauseSymbols = equiJoinClauses.stream()
+                        .map(JoinNode.EquiJoinClause::getRight)
+                        .collect(toImmutableSet());
+                ImmutableMap.Builder<String, Symbol> dynamicFiltersBuilder = ImmutableMap.builder();
+                node.getDynamicFilters().entrySet().stream()
+                        .filter(entry -> equiJoinClauseSymbols.contains(entry.getValue()))
+                        .forEach(dynamicFiltersBuilder::put);
+
+                Set<Symbol> existingAssignmentSymbols = ImmutableSet.copyOf(node.getDynamicFilters().values());
+                List<JoinNode.EquiJoinClause> unprocessedClauses = equiJoinClauses.stream()
+                        .filter(c -> !existingAssignmentSymbols.contains(c.getRight()))
+                        .collect(toImmutableList());
+
+                ImmutableList.Builder<Expression> predicatesBuilder = ImmutableList.builder();
+                for (JoinNode.EquiJoinClause clause : unprocessedClauses) {
+                    Symbol probeSymbol = clause.getLeft();
+                    Symbol buildSymbol = clause.getRight();
+                    String id = idAllocator.getNextId().toString();
+                    predicatesBuilder.add(createDynamicFilterExpression(id, probeSymbol.toSymbolReference()));
+                    dynamicFiltersBuilder.put(id, buildSymbol);
+                }
+                dynamicFilters = dynamicFiltersBuilder.build();
+                predicates = predicatesBuilder.build();
+            }
+
             if (leftSource != node.getLeft() ||
                     rightSource != node.getRight() ||
                     !filtersEquivalent ||
+                    !dynamicFilters.equals(node.getDynamicFilters()) ||
                     !ImmutableSet.copyOf(equiJoinClauses).equals(ImmutableSet.copyOf(node.getCriteria()))) {
                 leftSource = new ProjectNode(idAllocator.getNextId(), leftSource, leftProjections.build());
                 rightSource = new ProjectNode(idAllocator.getNextId(), rightSource, rightProjections.build());
 
+                if (!predicates.isEmpty()) {
```

Comment:
this shouldn't be needed. You should merge dynamic predicates with `leftPredicate` before:
```
 PlanNode leftSource = context.rewrite(node.getLeft(), leftPredicate);
```
this line. Otherwise it will take many executions of `PredicatePushDown` to push predicates.

Reply:
We are looking at `equiJoinClauses` for creating dynamic predicates and `equiJoinClauses` is created after the left side has already been rewritten.

Reply:
Sure, but you can move:
```
            PlanNode leftSource = context.rewrite(node.getLeft(), leftPredicate);
            PlanNode rightSource = context.rewrite(node.getRight(), rightPredicate);
```
after new equi clauses are created.